### PR TITLE
feat: use HuggingFace datasets for memory-mapped data loading

### PR DESCRIPTION
**Pull request by shawntsai**
## Problem

The current `VLMDataset` loads the entire parquet file into a PyArrow Table in RAM:
```python
self.table = pa.Table.from_batches(pq.ParquetFile(parquet_path).iter_batches())
```

With `num_workers=8` in the DataLoader, each worker forks and copies this table:
- pretrain_i2t.parquet: 4.1GB × 8 workers = **~33GB RAM needed**
- On a g6.2xlarge (30GB RAM): **OOM crash**

Users must reduce `num_workers` to 2, making training slower due to data loading bottleneck.

## Solution

Replace with HuggingFace `datasets.Dataset.from_parquet()` which provides **memory-mapped Arrow access**:
```python
from datasets import Dataset as HFDataset
self.dataset = HFDataset.from_parquet(parquet_path)
```

Memory-mapped files are shared across all DataLoader workers via the OS page cache — no per-worker copies.

## Results

| Metric | Before (PyArrow Table) | After (HF Datasets) |
|--------|----------------------|--------------------|
| RAM per worker | ~4.1GB (full copy) | Shared via mmap |
| 8 workers total RAM | ~33GB (OOM) | ~4-5GB |
| First load time | ~instant (but copies on fork) | ~76s (builds Arrow cache) |
| Subsequent loads | same | **<1s** (cached) |
| Process RSS | 4.1GB+ | **2.1GB** |

## Testing

Verified full pipeline on g6.2xlarge (NVIDIA L4, 30GB RAM):
- ✅ Dataset loads correctly (1,274,698 samples)
- ✅ DataLoader with `num_workers=4` produces valid batches
- ✅ Forward pass + loss computation works
- ✅ Shapes: input_ids (B,T), labels (B,T), pixel_values (B,1,3,256,256)

## Changes

- Replace `pyarrow` direct table loading with `datasets.Dataset.from_parquet()`
- Update `__getitem__` to use dict-style access instead of `.as_py()`
- Update `__main__` visualization to use HF datasets
- Remove unused `pyarrow` imports

## New dependency

`datasets` (HuggingFace) — already commonly installed alongside `transformers`.

### CODE REVIEW

**Comment by jingyaogong** [comment]
Thanks for the detailed benchmark. The memory issue is real: the current dataset keeps the parquet table in memory, and a high num_workers setting can put pressure on RAM-limited machines.

For now, I prefer to keep the dataset implementation simple and reduce the default num_workers from 8 to 2. MiniMind-V is meant to stay easy to inspect and run, and the HF datasets path introduces an additional cache/mmap layer with a longer first-load time. Users who train on larger machines can still increase --num_workers manually.

I may revisit mmap-based loading later if data loading becomes the main bottleneck.
